---
title: Analysis Formats
parent: File Formats
has_children: false
nav_order: 50
published: true
---

### Analysis read inputs (various)

The inputs to agFree variant analysis, i.e., pipeline `af analyze`,
are:
- genome files communicated via options `--genome` and `--genome-dir`
- name-sorted BAM files generated by `af align genome|genotype`
- indexed RE site files generated by `af locate genome|genotype`, `af genome index`, or `af genotype construct` and communicated via options `--re-sites-dir` and `--enzyme-name`

The required name-sorted BAM file is not communicated by a specific option.
Rather, it is read from the directory to which analysis output will also be written, i.e.,
`<--output-dir>/<--data-name>`.

### Targeted analysis regions (BED)

In some cases, your library may have been subjected to pre-sequencing target capture
or ONT adaptive sampling to enrich for target regions. Such target regions are communicated via option `--targets-bed-file`, which should point to a BED file with at least columns:
- chrom = RNAME of the chromosome
- start = 0-referenced start position
- end = 1-referenced end position

Leave `--targets-bed-file` as the default `NA` if no target regions were used, i.e.,
if whole genome sequencing (WGS) was performed.

### Fragment analysis outputs

The first step in all agFree variant analysis is to filter and index the read alignments against various quality metrics. Filtered alignments then PENDING.

### Structural variant (SV) analysis and filtering outputs

PENDING

### Single-nucleotide variant (SNV) analysis and filtering outputs

PENDING
